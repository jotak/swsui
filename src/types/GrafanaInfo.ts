export interface GrafanaInfo {
  dashboards: GrafanaDashboardInfo[];
}

interface GrafanaDashboardInfo {
  url: string;
  name: string;
  variables: GrafanaVariablesConfig;
}

interface GrafanaVariablesConfig {
  app?: string;
  namespace?: string;
  service?: string;
  workload?: string;
}
