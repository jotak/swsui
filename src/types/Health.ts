import { PfColors } from '../components/Pf/PfColors';
import { getName } from '../utils/RateIntervals';

interface HealthItem {
  status: Status;
  title: string;
  text?: string;
  children?: HealthSubItem[];
  report?: string;
}

interface HealthSubItem {
  status: Status;
  text: string;
}

export interface WorkloadStatus {
  name: string;
  replicas: number;
  available: number;
}

export interface RequestHealth {
  errorRatio: number;
  inboundErrorRatio: number;
  outboundErrorRatio: number;
}

export interface Status {
  name: string;
  color: string;
  priority: number;
  icon?: string;
  text?: string;
}

export const FAILURE: Status = {
  name: 'Failure',
  color: PfColors.Red100,
  priority: 3,
  icon: 'error-circle-o'
};
export const DEGRADED: Status = {
  name: 'Degraded',
  color: PfColors.Orange400,
  priority: 2,
  icon: 'warning-triangle-o'
};
export const HEALTHY: Status = {
  name: 'Healthy',
  color: PfColors.Green400,
  priority: 1,
  icon: 'ok'
};
export const NA: Status = {
  name: 'No health information',
  color: PfColors.Gray,
  priority: 0,
  text: 'N/A'
};

interface Thresholds {
  degraded: number;
  failure: number;
  unit: string;
}

export const REQUESTS_THRESHOLDS: Thresholds = {
  degraded: 0.1,
  failure: 20,
  unit: '%'
};

interface ThresholdStatus {
  value: number;
  status: Status;
  violation?: string;
}

// Use -1 rather than NaN to allow straigthforward comparison
const RATIO_NA = -1;

export const ratioCheck = (valid: number, total: number): Status => {
  if (total === 0) {
    return NA;
  } else if (valid === 0) {
    return FAILURE;
  } else if (valid === total) {
    return HEALTHY;
  }
  return DEGRADED;
};

export const mergeStatus = (s1: Status, s2: Status): Status => {
  return s1.priority > s2.priority ? s1 : s2;
};

const ascendingThresholdCheck = (value: number, thresholds: Thresholds): ThresholdStatus => {
  if (value >= thresholds.failure) {
    return {
      value: value,
      status: FAILURE,
      violation: value.toFixed(2) + thresholds.unit + '>=' + thresholds.failure + thresholds.unit
    };
  } else if (value >= thresholds.degraded) {
    return {
      value: value,
      status: DEGRADED,
      violation: value.toFixed(2) + thresholds.unit + '>=' + thresholds.degraded + thresholds.unit
    };
  }
  return { value: value, status: HEALTHY };
};

export const getRequestErrorsStatus = (ratio: number): ThresholdStatus => {
  if (ratio < 0) {
    return {
      value: RATIO_NA,
      status: NA
    };
  }
  return ascendingThresholdCheck(100 * ratio, REQUESTS_THRESHOLDS);
};

export const getRequestErrorsSubItem = (thresholdStatus: ThresholdStatus, prefix: string): HealthSubItem => {
  return {
    status: thresholdStatus.status,
    text: prefix + ': ' + (thresholdStatus.status === NA ? 'No requests' : thresholdStatus.value.toFixed(2) + '%')
  };
};

export const getRequestErrorsViolations = (reqIn: ThresholdStatus, reqOut: ThresholdStatus): string => {
  const violations: string[] = [];
  if (reqIn.violation) {
    violations.push(`Inbound errors ${reqIn.status.name.toLowerCase()}: ${reqIn.violation}`);
  }
  if (reqOut.violation) {
    violations.push(`Outbound errors ${reqOut.status.name.toLowerCase()}: ${reqOut.violation}`);
  }
  return violations.join(', ');
};

export abstract class Health {
  items: HealthItem[];

  constructor(items: HealthItem[]) {
    this.items = items;
  }

  getGlobalStatus(): Status {
    return this.items.map(i => i.status).reduce((prev, cur) => mergeStatus(prev, cur), NA);
  }

  getReport(): string[] {
    return this.items.filter(i => i.report !== undefined).map(i => i.report!);
  }
}

export class ServiceHealth extends Health {
  public static fromJson = (json: any, rateInterval: number) => new ServiceHealth(json.requests, rateInterval);

  private static computeItems(requests: RequestHealth, rateInterval: number): HealthItem[] {
    const items: HealthItem[] = [];
    {
      // Request errors
      const reqErrorsRatio = getRequestErrorsStatus(requests.errorRatio);
      const reqErrorsText = reqErrorsRatio.status === NA ? 'No requests' : reqErrorsRatio.value.toFixed(2) + '%';
      const item: HealthItem = {
        title: 'Error Rate',
        status: reqErrorsRatio.status,
        text: reqErrorsText + ' over ' + getName(rateInterval).toLowerCase()
      };
      if (reqErrorsRatio.violation) {
        item.report = `Error rate ${reqErrorsRatio.status.name.toLowerCase()}: ${reqErrorsRatio.violation}`;
      }
      items.push(item);
    }
    return items;
  }

  constructor(public requests: RequestHealth, public rateInterval: number) {
    super(ServiceHealth.computeItems(requests, rateInterval));
  }
}

export class AppHealth extends Health {
  public static fromJson = (json: any, rateInterval: number) =>
    new AppHealth(json.workloadStatuses, json.requests, rateInterval);

  private static computeItems(
    workloadStatuses: WorkloadStatus[],
    requests: RequestHealth,
    rateInterval: number
  ): HealthItem[] {
    const items: HealthItem[] = [];
    {
      // Pods
      let countInactive = 0;
      const children: HealthSubItem[] = workloadStatuses.map(d => {
        const status = ratioCheck(d.available, d.replicas);
        if (status === NA) {
          countInactive++;
        }
        return {
          text: d.name + ': ' + d.available + ' / ' + d.replicas,
          status: status
        };
      });
      const workloadStatus = children.map(i => i.status).reduce((prev, cur) => mergeStatus(prev, cur), NA);
      const item: HealthItem = {
        title: 'Workload Status',
        status: workloadStatus,
        children: children
      };
      if (countInactive > 0 && countInactive === workloadStatuses.length) {
        // No active deployment => special case for failure
        item.report = 'No active workload!';
        item.status = FAILURE;
      } else if (workloadStatus === FAILURE || workloadStatus === DEGRADED) {
        item.report = 'Pod workload ' + workloadStatus.name.toLowerCase();
      } else if (countInactive === 1) {
        item.report = 'One inactive workload';
      } else if (countInactive > 1) {
        item.report = `${countInactive} inactive workloads`;
      }
      items.push(item);
    }
    {
      // Request errors
      const reqIn = getRequestErrorsStatus(requests.inboundErrorRatio);
      const reqOut = getRequestErrorsStatus(requests.outboundErrorRatio);
      const both = mergeStatus(reqIn.status, reqOut.status);
      const item: HealthItem = {
        title: 'Error Rate over ' + getName(rateInterval).toLowerCase(),
        status: both,
        children: [getRequestErrorsSubItem(reqIn, 'Inbound'), getRequestErrorsSubItem(reqOut, 'Outbound')]
      };
      const violations = getRequestErrorsViolations(reqIn, reqOut);
      if (violations.length > 0) {
        item.report = violations;
      }
      items.push(item);
    }
    return items;
  }

  constructor(public workloadStatuses: WorkloadStatus[], public requests: RequestHealth, public rateInterval: number) {
    super(AppHealth.computeItems(workloadStatuses, requests, rateInterval));
  }
}

export class WorkloadHealth extends Health {
  public static fromJson = (json: any, rateInterval: number) =>
    new WorkloadHealth(json.workloadStatus, json.requests, rateInterval);

  private static computeItems(
    workloadStatus: WorkloadStatus,
    requests: RequestHealth,
    rateInterval: number
  ): HealthItem[] {
    const items: HealthItem[] = [];
    {
      // Pods
      const workStatus = ratioCheck(workloadStatus.available, workloadStatus.replicas);
      const item: HealthItem = {
        title: 'Workloads Status',
        status: workStatus,
        text: String(workloadStatus.available + ' / ' + workloadStatus.replicas)
      };
      if (workStatus === FAILURE || workStatus === DEGRADED) {
        item.report = 'Pod workload ' + workStatus.name.toLowerCase();
      }
      items.push(item);
    }
    {
      // Request errors
      const reqIn = getRequestErrorsStatus(requests.inboundErrorRatio);
      const reqOut = getRequestErrorsStatus(requests.outboundErrorRatio);
      const both = mergeStatus(reqIn.status, reqOut.status);
      const item: HealthItem = {
        title: 'Error Rate over ' + getName(rateInterval).toLowerCase(),
        status: both,
        children: [getRequestErrorsSubItem(reqIn, 'Inbound'), getRequestErrorsSubItem(reqOut, 'Outbound')]
      };
      const violations = getRequestErrorsViolations(reqIn, reqOut);
      if (violations.length > 0) {
        item.report = violations;
      }
      items.push(item);
    }
    return items;
  }

  constructor(public workloadStatus: WorkloadStatus, public requests: RequestHealth, public rateInterval: number) {
    super(WorkloadHealth.computeItems(workloadStatus, requests, rateInterval));
  }
}

export const healthNotAvailable = (): AppHealth => {
  return new AppHealth([], { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 }, 60);
};

export type NamespaceAppHealth = { [app: string]: AppHealth };
export type NamespaceServiceHealth = { [service: string]: ServiceHealth };
export type NamespaceWorkloadHealth = { [workload: string]: WorkloadHealth };
