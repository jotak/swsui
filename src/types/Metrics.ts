export interface Metrics {
  metrics: { [key: string]: MetricGroup };
  histograms: { [key: string]: Histogram };
}

export type Histogram = { [key: string]: MetricGroup };

export interface MetricGroup {
  matrix: TimeSeries[];
}

export type Metric = {
  [key: string]: string;
};

export interface TimeSeries {
  metric: Metric;
  values: Datapoint[];
  name: string;
}

// First is timestamp, second is value
export type Datapoint = [number, number];

export enum MetricsObjectTypes {
  SERVICE,
  WORKLOAD,
  APP
}

export interface MetricsStatsResult {
  // Key is built from query params, see StatsComparison.genKey. The same key needs to be generated server-side for matching.
  stats: { [key: string]: MetricsStats };
  // Note: warnings here is for non-blocking errors, it's set when some stats are available, but not all, for instance due to inaccessible namespaces
  // For more serious errors (e.g. prometheus inaccessible) the query would return an HTTP error
  warnings?: string[];
}

export interface MetricsStats {
  responseTimes: Stat[];
}

export interface Stat {
  name: string;
  value: number;
}
