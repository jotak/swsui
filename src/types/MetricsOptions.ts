import { MetricsQuery } from '@kiali/k-charted-pf3';

export interface IstioMetricsOptions extends MetricsQuery {
  direction: Direction;
  filters?: string[];
  requestProtocol?: string;
  reporter: Reporter;
}

export type Reporter = 'source' | 'destination';
export type Direction = 'inbound' | 'outbound';
