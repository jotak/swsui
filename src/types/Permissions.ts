export interface ResourcePermissions {
  update: boolean;
  delete: boolean;
}
