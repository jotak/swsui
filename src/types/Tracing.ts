export type TracingQuery = {
  startMicros: number;
  endMicros?: number;
  // TODO: tags?
  // tags?: {[key: string]: string}
};

export type Span = {
  traceID: string;
  spanID: string;
  operationName: string;
  startTime: number;
  duration: number;
  tags: Tag[];
  warnings?: string[];
  traceSize: number;
  // TODO: more meta-data? logs ...
};

export type Tag = {
  key: string;
  type: string;
  value: any;
};
