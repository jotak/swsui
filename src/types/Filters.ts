import { TextInputTypes } from '@patternfly/react-core';
import { HEALTHY, DEGRADED, FAILURE, NA, Health } from 'types/Health';

// FilterValue maps a Patternfly property. Modify with care.
export interface FilterValue {
  id: string;
  title: string;
}

enum NonInputTypes {
  typeAhead = 'typeahead',
  select = 'select',
  label = 'label',
  nsLabel = 'nsLabel'
}

export const FilterTypes = {
  ...TextInputTypes,
  ...NonInputTypes
};

type FilterTypes = NonInputTypes | TextInputTypes;

// FilterDefinition maps a Patternfly property. Modify with care.
export interface FilterDefinition {
  id: string;
  title: string;
  placeholder: string;
  filterType: FilterTypes;
  action: string;
  filterValues: FilterValue[];
  loader?: () => Promise<FilterValue[]>;
}

export type Filter<T> = FilterDefinition & {
  check: (item: T, active: ActiveFiltersInfo) => boolean;
};

export type PromiseFilter<T> = FilterDefinition & {
  checkAsync: (item: T, active: ActiveFiltersInfo) => Promise<boolean>;
};

export const promiseFilter = <T>(filter: Filter<T> | PromiseFilter<T>): PromiseFilter<T> => {
  return (filter as PromiseFilter<T>).checkAsync !== undefined
    ? (filter as PromiseFilter<T>)
    : { ...filter, checkAsync: (items, active) => Promise.resolve((filter as Filter<T>).check(items, active)) };
};

export const FILTER_ACTION_APPEND = 'append';
export const FILTER_ACTION_UPDATE = 'update';

export interface ActiveFilter {
  id: string;
  title: string;
  value: string;
}

export type LabelOperation = 'and' | 'or';
export const ID_LABEL_OPERATION = 'opLabel';
export const DEFAULT_LABEL_OPERATION: LabelOperation = 'or';

export interface ActiveFiltersInfo {
  filters: ActiveFilter[];
  op: LabelOperation;
}

///////////////////
// Some constants

export const presenceValues: FilterValue[] = [
  {
    id: 'present',
    title: 'Present'
  },
  {
    id: 'notpresent',
    title: 'Not Present'
  }
];

export const presenceCheck = (filter: ActiveFilter, value: boolean) => {
  return filter.value === (value ? 'Present' : 'Not Present');
};

export const istioSidecarFilter: Filter<{ istioSidecar: boolean }> = {
  id: 'istiosidecar',
  title: 'Istio Sidecar',
  placeholder: 'Filter by IstioSidecar Validation',
  filterType: FilterTypes.select,
  action: FILTER_ACTION_UPDATE,
  filterValues: presenceValues,
  check: (item, active) => active.filters.some(f => presenceCheck(f, item.istioSidecar))
};

export const healthFilter: PromiseFilter<{ healthPromise: Promise<Health> }> = {
  id: 'health',
  title: 'Health',
  placeholder: 'Filter by Health',
  filterType: FilterTypes.select,
  action: FILTER_ACTION_APPEND,
  filterValues: [
    {
      id: HEALTHY.name,
      title: HEALTHY.name
    },
    {
      id: DEGRADED.name,
      title: DEGRADED.name
    },
    {
      id: FAILURE.name,
      title: FAILURE.name
    },
    {
      id: 'na',
      title: NA.name
    }
  ],
  checkAsync: (item, active) =>
    item.healthPromise.then(h => active.filters.some(f => f.value === h.getGlobalStatus().name))
};
