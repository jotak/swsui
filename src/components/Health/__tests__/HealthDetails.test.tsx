import * as React from 'react';
import { shallow } from 'enzyme';

import { HealthDetails } from '../HealthDetails';
import { ServiceHealth } from '../../../types/Health';

describe('HealthDetails', () => {
  it('renders healthy', () => {
    const health = new ServiceHealth(
      { inbound: { healthy: 1, total: 1 }, outbound: { healthy: 1, total: 1 } },
      { errorRatio: -1 },
      60
    );

    let wrapper = shallow(<HealthDetails id="svc" health={health} headline="" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders envoy degraded', () => {
    const health = new ServiceHealth(
      { inbound: { healthy: 1, total: 10 }, outbound: { healthy: 1, total: 1 } },
      { errorRatio: -1 },
      60
    );

    let wrapper = shallow(<HealthDetails id="svc" health={health} headline="" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders deployments failure', () => {
    const health = new ServiceHealth(
      { inbound: { healthy: 1, total: 10 }, outbound: { healthy: 1, total: 1 } },
      { errorRatio: -1 },
      60
    );

    let wrapper = shallow(<HealthDetails id="svc" health={health} headline="" />);
    expect(wrapper).toMatchSnapshot();
  });
});
