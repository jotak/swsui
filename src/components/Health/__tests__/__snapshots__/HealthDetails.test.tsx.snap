// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthDetails renders deployments failure 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HealthDetails
    headline=""
    health={
      ServiceHealth {
        "items": Array [
          Object {
            "status": Object {
              "color": "#72767b",
              "name": "No health information",
              "priority": 0,
              "text": "N/A",
            },
            "text": "No requests over last min",
            "title": "Error Rate",
          },
        ],
        "rateInterval": 60,
        "requests": Object {
          "errorRatio": -1,
          "inboundErrorRatio": -1,
          "outboundErrorRatio": -1,
        },
      }
    }
    id="svc"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "defaultOverlayShown": false,
      "overlay": <Popover
        bsClass="popover"
        id="svc-health-tooltip"
        placement="right"
        title=""
      >
        <div>
          <strong>
            <span
              style={
                Object {
                  "color": "#72767b",
                }
              }
            >
              N/A
            </span>
             Error Rate: 
          </strong>
          No requests over last min
        </div>
      </Popover>,
      "placement": "right",
      "rootClose": false,
      "trigger": Array [
        "hover",
        "focus",
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "defaultOverlayShown": false,
        "overlay": <Popover
          bsClass="popover"
          id="svc-health-tooltip"
          placement="right"
          title=""
        >
          <div>
            <strong>
              <span
                style={
                  Object {
                    "color": "#72767b",
                  }
                }
              >
                N/A
              </span>
               Error Rate: 
            </strong>
            No requests over last min
          </div>
        </Popover>,
        "placement": "right",
        "rootClose": false,
        "trigger": Array [
          "hover",
          "focus",
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`HealthDetails renders healthy 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HealthDetails
    headline=""
    health={
      ServiceHealth {
        "items": Array [
          Object {
            "status": Object {
              "color": "#72767b",
              "name": "No health information",
              "priority": 0,
              "text": "N/A",
            },
            "text": "No requests over last min",
            "title": "Error Rate",
          },
        ],
        "rateInterval": 60,
        "requests": Object {
          "errorRatio": -1,
          "inboundErrorRatio": -1,
          "outboundErrorRatio": -1,
        },
      }
    }
    id="svc"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "defaultOverlayShown": false,
      "overlay": <Popover
        bsClass="popover"
        id="svc-health-tooltip"
        placement="right"
        title=""
      >
        <div>
          <strong>
            <span
              style={
                Object {
                  "color": "#72767b",
                }
              }
            >
              N/A
            </span>
             Error Rate: 
          </strong>
          No requests over last min
        </div>
      </Popover>,
      "placement": "right",
      "rootClose": false,
      "trigger": Array [
        "hover",
        "focus",
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "defaultOverlayShown": false,
        "overlay": <Popover
          bsClass="popover"
          id="svc-health-tooltip"
          placement="right"
          title=""
        >
          <div>
            <strong>
              <span
                style={
                  Object {
                    "color": "#72767b",
                  }
                }
              >
                N/A
              </span>
               Error Rate: 
            </strong>
            No requests over last min
          </div>
        </Popover>,
        "placement": "right",
        "rootClose": false,
        "trigger": Array [
          "hover",
          "focus",
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
